File: store/shortcut.go
package store

import (
	"context"

	storepb "github.com/bshort/monotreme/proto/gen/store"
)

type UpdateShortcut struct {
	ID int32

	Name              *string
	Link              *string
	Title             *string
	Description       *string
	Visibility        *storepb.Visibility
	Tag               *string
	OpenGraphMetadata *storepb.OpenGraphMetadata
}

type FindShortcut struct {
	ID             *int32
	CreatorID      *int32
	Name           *string
	VisibilityList []storepb.Visibility
	Tag            *string
}

type DeleteShortcut struct {
	ID int32
}

func (s *Store) CreateShortcut(ctx context.Context, create *storepb.Shortcut) (*storepb.Shortcut, error) {
	shortcut, err := s.driver.CreateShortcut(ctx, create)
	if err != nil {
		return nil, err
	}
	s.shortcutCache.Store(shortcut.Id, shortcut)
	return shortcut, nil
}

func (s *Store) UpdateShortcut(ctx context.Context, update *UpdateShortcut) (*storepb.Shortcut, error) {
	shortcut, err := s.driver.UpdateShortcut(ctx, update)
	if err != nil {
		return nil, err
	}
	s.shortcutCache.Store(shortcut.Id, shortcut)
	return shortcut, nil
}

func (s *Store) ListShortcuts(ctx context.Context, find *FindShortcut) ([]*storepb.Shortcut, error) {
	list, err := s.driver.ListShortcuts(ctx, find)
	if err != nil {
		return nil, err
	}
	for _, shortcut := range list {
		s.shortcutCache.Store(shortcut.Id, shortcut)
	}
	return list, nil
}

func (s *Store) GetShortcut(ctx context.Context, find *FindShortcut) (*storepb.Shortcut, error) {
	if find.ID != nil {
		if cache, ok := s.shortcutCache.Load(*find.ID); ok {
			if shortcut, ok := cache.(*storepb.Shortcut); ok {
				return shortcut, nil
			}
		}
	}

	shortcuts, err := s.ListShortcuts(ctx, find)
	if err != nil {
		return nil, err
	}

	if len(shortcuts) == 0 {
		return nil, nil
	}

	shortcut := shortcuts[0]
	s.shortcutCache.Store(shortcut.Id, shortcut)
	return shortcut, nil
}

func (s *Store) DeleteShortcut(ctx context.Context, delete *DeleteShortcut) error {
	if err := s.driver.DeleteShortcut(ctx, delete); err != nil {
		return err
	}

	s.shortcutCache.Delete(delete.ID)
	return nil
}

File: frontend/web/src/stores/v1/shortcut.ts
import { isEqual } from "lodash-es";
import { create } from "zustand";
import { shortcutServiceClient } from "@/grpcweb";
import { Shortcut } from "@/types/proto/api/v2/shortcut_service";

interface ShortcutState {
  shortcutMapByName: Record<string, Shortcut>;
  fetchShortcutList: () => Promise<Shortcut[]>;
  fetchShortcutById: (id: number) => Promise<Shortcut>;
  getOrFetchShortcutByName: (name: string) => Promise<Shortcut>;
  getShortcutByName: (name: string) => Shortcut;
  getShortcutList: () => Shortcut[];
  createShortcut: (shortcut: Shortcut) => Promise<Shortcut>;
  updateShortcut: (shortcut: Partial<Shortcut>, updateMask: string[]) => Promise<Shortcut>;
  deleteShortcut: (name: string) => Promise<void>;
}

const useShortcutStore = create<ShortcutState>()((set, get) => ({
  shortcutMapById: {},
  shortcutMapByName: {},
  fetchShortcutList: async () => {
    const { shortcuts } = await shortcutServiceClient.listShortcuts({});
    const shortcutMap = get().shortcutMapByName;
    shortcuts.forEach((shortcut) => {
      shortcutMap[shortcut.name] = shortcut;
    });
    set(shortcutMap);
    return shortcuts;
  },
  fetchShortcutById: async (id: number) => {
    const { shortcut } = await shortcutServiceClient.getShortcutById({
      id: id,
    });
    if (!shortcut) {
      throw new Error(`Shortcut with id ${id} not found`);
    }
    return shortcut;
  },
  getOrFetchShortcutByName: async (name: string) => {
    const shortcutMap = get().shortcutMapByName;
    if (shortcutMap[name]) {
      return shortcutMap[name] as Shortcut;
    }

    const { shortcut } = await shortcutServiceClient.getShortcut({
      name,
    });
    if (!shortcut) {
      throw new Error(`Shortcut with name ${name} not found`);
    }

    shortcutMap[name] = shortcut;
    set(shortcutMap);
    return shortcut;
  },
  getShortcutByName: (name: string) => {
    const shortcutMap = get().shortcutMapByName;
    return shortcutMap[name] || unknownShortcut;
  },
  getShortcutList: () => {
    return Object.values(get().shortcutMapByName);
  },
  createShortcut: async (shortcut: Shortcut) => {
    const { shortcut: createdShortcut } = await shortcutServiceClient.createShortcut({
      shortcut: shortcut,
    });
    if (!createdShortcut) {
      throw new Error(`Failed to create shortcut`);
    }
    const shortcutMap = get().shortcutMapByName;
    shortcutMap[createdShortcut.name] = createdShortcut;
    set(shortcutMap);
    return createdShortcut;
  },
  updateShortcut: async (shortcut: Partial<Shortcut>, updateMask: string[]) => {
    const { shortcut: updatedShortcut } = await shortcutServiceClient.updateShortcut({
      shortcut: shortcut,
      updateMask,
    });
    if (!updatedShortcut) {
      throw new Error(`Failed to update shortcut`);
    }
    const shortcutMap = get().shortcutMapByName;
    shortcutMap[updatedShortcut.name] = updatedShortcut;
    set(shortcutMap);
    return updatedShortcut;
  },
  deleteShortcut: async (name: string) => {
    await shortcutServiceClient.deleteShortcut({
      name,
    });
    const shortcutMap = get().shortcutMapByName;
    delete shortcutMap[name];
    set(shortcutMap);
  },
}));

const unknownShortcut: Shortcut = Shortcut.fromPartial({
  id: -1,
  name: "Unknown",
});

export const getShortcutUpdateMask = (shortcut: Shortcut, updatingShortcut: Shortcut) => {
  const updateMask: string[] = [];
  if (!isEqual(shortcut.name, updatingShortcut.name)) {
    updateMask.push("name");
  }
  if (!isEqual(shortcut.link, updatingShortcut.link)) {
    updateMask.push("link");
  }
  if (!isEqual(shortcut.title, updatingShortcut.title)) {
    updateMask.push("title");
  }
  if (!isEqual(shortcut.description, updatingShortcut.description)) {
    updateMask.push("description");
  }
  if (!isEqual(shortcut.tags, updatingShortcut.tags)) {
    updateMask.push("tags");
  }
  if (!isEqual(shortcut.visibility, updatingShortcut.visibility)) {
    updateMask.push("visibility");
  }
  if (!isEqual(shortcut.ogMetadata, updatingShortcut.ogMetadata)) {
    updateMask.push("og_metadata");
  }
  return updateMask;
};

export default useShortcutStore;
